server/lomas_server/dp_queries/dp_libraries/smartnoise_synth.py

```diff
@@ -0,0 +1,307 @@
+from typing import Dict, List
+
+import numpy as np
+import pandas as pd
+from snsynth import Synthesizer
+from snsynth.transform import (  # DateTimeTransformer,
+    AnonymizationTransformer,
+    BinTransformer,
+    ChainTransformer,
+    LabelTransformer,
+    MinMaxTransformer,
+    OneHotEncoder,
+)
+from snsynth.transform.table import TableTransformer
+
+from constants import (
+    SSYNTH_PRIVATE_COLUMN,
+    DPLibraries,
+    SSynthColumnType,
+    SSynthSynthesizer,
+    SSynthTableTransStyle,
+)
+from dp_queries.dp_libraries.utils import serialise_model
+from dp_queries.dp_querier import DPQuerier
+from utils.collection_models import Metadata
+from utils.error_handler import (
+    ExternalLibraryException,
+    InternalServerException,
+    InvalidQueryException,
+)
+from utils.query_models import SmartnoiseSynthModel
+
+
+class SmartnoiseSynthQuerier(DPQuerier):
+    """
+    Concrete implementation of the DPQuerier ABC for the SmartNoiseSynth library.
+    """
+
+    def cost(self, query_json: SmartnoiseSynthModel) -> tuple[float, float]:
+        """Return cost of query
+
+        Args:
+            query_json (SmartnoiseSynthModelCost): JSON request object for the query.
+
+        Returns:
+            tuple[float, float]: The tuple of costs, the first value
+                is the epsilon cost, the second value is the delta value.
+        """
+        # model.odometer.spent
+        return query_json.epsilon, query_json.delta
```

Comment:
Is this an upper bound on the cost? ie. is it possible that the model spends less ?

Reply:
neither :/ it can also spend more based on my understanding.
it is a placeholder for now